src/dmd/lexer.d

```diff
@@ -423,7 +423,26 @@ class Lexer
                     p += 3;
                 }
                 else
+                {
                     t.value = charConstant(t);
+                    if (t.unsvalue == '?' && errors())
```

Comment:
This is not a good test: '?' is a valid character and `errors()` might return true from unrelated previous errors. The recovery should better be in `charConstant` when the actual error is reported.

Reply:
`t.unsvalue` is set to '?' after an `error("unterminated character constant")` is thrown. This happens only in the `charConstant` function. For reference: https://github.com/dlang/dmd/blob/42cc4e81e8b03664293752db83568af74b4435ef/src/dmd/lexer.d#L1786 and https://github.com/dlang/dmd/blob/42cc4e81e8b03664293752db83568af74b4435ef/src/dmd/lexer.d#L1807

So whenever `t.unsvalue` is '?' because of an error, the `errors()` call will return true. This is the case I want to handle.

Reply:
'?' is a regular value here: 
https://github.com/dlang/dmd/blob/42cc4e81e8b03664293752db83568af74b4435ef/src/dmd/lexer.d#L1801
`errors()` can be true even before the call to `charConstant`.

Reply:
Oh, yes. You are right. I'll think about a better condition.

Reply:
You don't need a condition, just put it into a local function in charConstant when the error is emitted.

Reply:
Done, thanks.